### ISSUE TITLE
Listing projects on build service is excessively slow

### ISSUE THREAD

**Issue by AdamMajer** [openSUSE/open-build-service#11239]
I'm undecided if this a feature request or a bug.

```
> time osc ls / > /dev/null; time osc ls / > /dev/null;

real    0m2.027s
user    0m0.586s
sys     0m0.064s

real    0m1.809s
user    0m0.605s
sys     0m0.032s
```

Outside of the `osc` usage, it is often desired to query a list of projects and then query the packages of that project, etc. It *seems* this operation always hits the disk on the backend and results in this rather large 2s delay.

Instead of querying the state every single time, the backend (or maybe the API?) should cache the project list and use an alternative mechanism to watch for changes. RabbitMQ comes to mind.

**Comment by coolo** [comment]
This is purely SQL. The killer is mainly the XML rendering of the large document

**Comment by hennevogel** [comment]
It's 89353 objects.... not sure what you expect.

```
osc ls |wc -l            
89353
```

And not sure what you mean by querying for state, projects have no state. Maybe you want to tell us what you are actually trying to achieve?

**Comment by coolo** [comment]
Note that osc api /source > /dev/null takes 0.4 seconds less, even though it's the same API route. That's what it takes osc to parse the XML document. 

In essence: XML is just not a good transfer method for an API transferring a lot of data - but that ship has sailed.

**Comment by hennevogel** [comment]
This has little to do with transfer method or rendering engine. Requesting a JSON document over HTTP with 89353 entries in it will be slow too...

So @AdamMajer what is it you try to achieve?

**Comment by marcus-h** [comment]
On 2021-06-14 06:28:59 -0700, Stephan Kulow wrote:
> Note that osc api /source > /dev/null takes 0.4 seconds less, even though it's the same API route. That's what it takes osc to parse the XML document.
>
If there's a real need, we could improve that (instead of reading the whole
tree into memory we could do some SAX like/event-driven parsing...).

**Comment by AdamMajer** [comment]
```
time curl -b 'openSUSE_session='....' -o /dev/null https://api.opensuse.org/source

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 3413k    0 3413k    0     0  2380k      0 --:--:--  0:00:01 --:--:-- 2380k

real    0m1.662s
user    0m0.029s
sys     0m0.058s
```

The problem is not the size of the data. The problem is not the data encapsulation format (xml vs. json vs. plain text). The problem is not on the osc side parsing anything. The problem is that the API stalls for almost 2 seconds until the data is actually starts to be fetched. Every time it's being called, it's stalling.

This API call is actually used by services like the OBS plugin for Visual Studio Code to provide completion on typing. Maybe that's not so efficient bandwidth wise, but it's done once per instance (and gzip compressed would be 500kB). Anyway, I attach a packet capture timing of this stall.

![image](https://user-images.githubusercontent.com/1211498/121926549-5dc83380-cd3e-11eb-9aad-36ce80b869e4.png)

**Comment by AdamMajer** [comment]
> It's 89353 objects.... not sure what you expect.

I would expect network limited transfer that starts almost as fast as a request is issued.

> And not sure what you mean by querying for state, projects have no state. Maybe you want to tell us what you are actually trying to achieve?

I mean that every time this request is made, it *seems* like the backend is getting hit and an uncached listing of some directory(ies) is getting returned. I mean that the API seems to have no notion of what is the current state of the project list and is doing some heavy tasks in trying to come up with the list.

Project listing is kind of unavoidable part of doing autocompletion on typing. For some applications maybe it would be better to filter server side but for others it may make more sense to filter client side. It depends. The current project list size is kind of on the border between the two approaches.

**Comment by coolo** [comment]
No, the proper API for autocompletion is the search controller. You can still continue client side filtering once you have the first 3 chars server side filtered.

time osc api "/search/project/id?match=starts-with(@name,'home:c')"
real    0m0,603s
user    0m0,215s
sys     0m0,031s

**Comment by coolo** [comment]
And about your theories:
```
method=GET path=/source format=*/* controller=SourceController action=index status=200 duration=1516.38 view=997.27 db=242.01 params={} backend=0 user=adamm
```

The duration is 1.5s with the rails code, the xml creation is almost a full second and SQL adds another 242ms, the backend is not involved at all (nor any filesystem that isn't in in the DB)

**Comment by AdamMajer** [comment]
> The duration is 1.5s with the rails code, the xml creation is almost a full second and SQL adds another 242ms, the backend is not involved at all (nor any filesystem that isn't in in the DB)

I don't have answers here. It's clearly a heavy operation the way it runs now.

> No, the proper API for autocompletion is the search controller.

Thank you, that's a faster solution. But a complete list can be used for **more** than autocompletion - you can list all the projects before one starts to filter them, like when you look at https://build.opensuse.org/project . The data is queried once and can be used repeatedly in the application so there are advantaged to doing this.

Like I said, the size of the list is kind of on the cusp between searching on server side or fetching entire list for client side filtering and other purposes.

**Comment by coolo** [comment]
Well, serious APIs wouldn't even allow you to fetch the full list but limit it to say 1000 and offer you an offset to fetch the remains. Then each API request would be a dozen ms and you would sum up client side.

**Comment by AdamMajer** [comment]
I would be perfectly OK with that change to limit results to some reasonable number and have offset. The number should probably be the maximum size before we start growing processing time linearly. I would guess this would be around 10000 rows based on the response times from the search controller.

In addition, if we had a way of filtering home and maintenance projects without hitting the search controller, that would basically account for the entire usage of this endpoint.

**Comment by hennevogel** [comment]
Woulda, coulda, shoulda. What stops you from using `/search/project/id` for your use case?

**Comment by AdamMajer** [comment]
> Woulda, coulda, shoulda. What stops you from using `/search/project/id` for your use case?

At very **least** it would help that the documentation would point users in this direction then.

https://github.com/openSUSE/open-build-service/blob/master/docs/api/api/api.txt

> == Sources
>         === Projects
>         GET /source/
>           Read list of projects.
>         XmlResult: directory
>         Parameters:
>           deleted: show deleted projects instead of existing

vs.

> GET /search/project/id
>           Searches for project metadata analogous to /search/project, only the root
>           element is returned without any children.
>         Parameters:
>           match: xpath predicate, mandatory
>         XmlResult: collection

The 2nd one is not even clear what it does especially if someone is looking to read a list of projects on OBS.

Also, it appears just as slow (18k projects, with most Maintenance),

```
time osc api "https://api.opensuse.org//search/project/id?match=not(starts-with(@name,'home:'))"

real    0m1.572s
user    0m0.254s
sys     0m0.036s
```

I haven't figured out how to add `and` condition to remove also Maintenance. My xpath-foo not strong :cry:

**Comment by hennevogel** [comment]
> I haven't figured out how to add and condition to remove also Maintenance. My xpath-foo not strong cry

`+and+` you are not alone 😉  But there is [help...](https://www.w3schools.com/XML/xpath_intro.asp)

**Comment by coolo** [comment]
It's not a very complicated language - once you realize you need to quote for command line and HTTP :)

`osc api "https://api.opensuse.org//search/project/id?match=not(starts-with(@name,'home:'))+and+not(starts-with(@name,'openSUSE:')))"` is 0.8s

**Comment by hennevogel** [comment]
So what is left are a couple of feature requests:

- Adding offset/limit to more API endpoints, especially `/source`
- Mention `/search` in the `/source` documentation
